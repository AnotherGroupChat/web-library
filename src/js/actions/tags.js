'use strict';

import api from 'zotero-api-client';

const getApi = ({ config, libraryKey }, requestType, queryConfig) => {
	switch(requestType) {
		case 'TAGS_IN_COLLECTION':
			return api(config.apiKey, config.apiConfig)
				.library(libraryKey)
				.collections(queryConfig.collectionKey)
				.items()
				.top()
				.tags();
		case 'TAGS_IN_TRASH_ITEMS':
			return api(config.apiKey, config.apiConfig)
				.library(libraryKey)
				.items()
				.trash()
				.tags()
		case 'TAGS_IN_PUBLICATIONS_ITEMS':
			return api(config.apiKey, config.apiConfig)
				.library(libraryKey)
				.items()
				.publications()
				.tags()
		case 'TAGS_IN_ITEMS_BY_QUERY':
			var configuredApi = api(config.apiKey, config.apiConfig)
				.library(libraryKey)
				.items()
			if(queryConfig.collectionKey) {
				configuredApi = configuredApi.collections(queryConfig.collectionKey).top()
			} else if(queryConfig.isMyPublications) {
				configuredApi = configuredApi.publications();
			} else if(queryConfig.isTrash) {
				configuredApi = configuredApi.trash();
			} else {
				configuredApi = configuredApi.top();
			}
			return configuredApi.tags();
		case 'TAGS_FOR_ITEM':
			return api(config.apiKey, config.apiConfig)
				.library(libraryKey)
				.items(queryConfig.itemKey)
				.top()
				.tags();
		case 'TAGS_IN_TOP_ITEMS':
			return api(config.apiKey, config.apiConfig)
				.library(libraryKey)
				.items()
				.top()
				.tags();
		default:
		case 'TAGS_IN_LIBRARY':
			return api(config.apiKey, config.apiConfig)
				.library(libraryKey)
				.tags();
	}
}

const fetchTags = (type, queryConfig, queryOptions = {}) => {
	return async (dispatch, getState) => {
		const state = getState();
		const config = state.config;
		const { libraryKey } = state.current;
		const api = getApi({ config, libraryKey }, type, queryConfig);

		dispatch({
			type: `REQUEST_${type}`,
			libraryKey, ...queryConfig, ...queryOptions
		});

		try {
			const response = await api.get(queryOptions);
			const tags = response.getData().map((tagData, index) => ({
				tag: tagData.tag,
				[Symbol.for('meta')]: response.getMeta()[index] || {}
			}));
			const totalResults = parseInt(response.response.headers.get('Total-Results'), 10);

			dispatch({
				type: `RECEIVE_${type}`,
				libraryKey, tags, response, totalResults,
				...queryConfig,...queryOptions
			});

			return tags;
		} catch(error) {
			dispatch({
				type: `ERROR_${type}`,
				libraryKey, error, ...queryConfig, ...queryOptions
			});

			throw error;
		}
	}
}

const fetchTagsInCollection = (collectionKey, queryOptions) => {
	return fetchTags('TAGS_IN_COLLECTION', { collectionKey }, queryOptions);
};

const fetchTagsInLibrary = queryOptions => {
	return fetchTags('TAGS_IN_LIBRARY', { }, queryOptions);
};

const fetchTagsForItem = (itemKey, queryOptions) => {
	return fetchTags('TAGS_FOR_ITEM', { itemKey }, queryOptions);
};

const fetchTagsForTrashItems = queryOptions => {
	return fetchTags('TAGS_IN_TRASH_ITEMS', {}, queryOptions);
};

const fetchTagsForPublicationsItems = queryOptions => {
	return fetchTags('TAGS_IN_PUBLICATIONS_ITEMS', {}, queryOptions);
};

const fetchTagsForTopItems = queryOptions => {
	return fetchTags('TAGS_IN_TOP_ITEMS', { }, queryOptions);
};

const fetchTagsForItemsByQuery = (query, queryOptions) => {
	const { collectionKey = null, itemTag = null, itemQ = null, itemQMode = null,
		isTrash, isMyPublications } = query;
	const queryConfig = { collectionKey, isTrash, isMyPublications };

	return fetchTags(
		'TAGS_IN_ITEMS_BY_QUERY', queryConfig, { ...queryOptions, itemTag, itemQ, itemQMode }
	);
}

export {
	fetchTagsForItem,
	fetchTagsForItemsByQuery,
	fetchTagsForPublicationsItems,
	fetchTagsForTopItems,
	fetchTagsForTrashItems,
	fetchTagsInCollection,
	fetchTagsInLibrary,
};
