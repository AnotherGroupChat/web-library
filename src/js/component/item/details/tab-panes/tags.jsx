'use strict';

import React from 'react';
import cx from 'classnames';
import Tags from '../../../tags';

class TagsTabPane extends React.PureComponent {
	render() {
		return (
			<div className={ cx({
				'tab-pane': true,
				'tags': true,
				'active': this.props.isActive
			}) }>
				<h5 className="h2 tab-pane-heading">Tags</h5>
				<Tags
					item={ this.props.item }
					tags={ this.props.item.tags }
					isProcessingTags={ this.props.isProcessingTags }
					isReadOnly={ this.props.isReadOnly }
					onAddTag={ this.props.onAddTag }
					onDeleteTag={ this.props.onDeleteTag }
					onUpdateTag={ this.props.onUpdateTag }
				/>
			</div>
		);
	}
}

export default TagsTabPane;
