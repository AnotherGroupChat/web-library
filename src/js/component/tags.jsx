import React, { useState } from 'react';
import PropTypes from 'prop-types';

import Button from './ui/button';
import Editable from './editable';
import Icon from './ui/icon';
import { Toolbar, ToolGroup } from './ui/toolbars';
import { deduplicateByKey, sortByKey } from '../utils';
import { pick } from '../common/immutable';

var nextId = 0;

const Tags = ({ tagColors, itemKey, tags: initalTags, isReadOnly, updateItem }) => {
	const [tags, setTags] = useState(initalTags.map(t => ({ ...t, id: ++nextId })));
	const [tagRedacted, setTagRedacted] = useState(null);

	const handleCommit = async (newTagValue, hasChanged, ev) => {
		const tag = ev.currentTarget.closest('[data-tag]').dataset.tag;
		setTagRedacted(null);

		if(!hasChanged) {
			setTags(tags.filter(t => t.tag !== ''));
			return;
		}

		const updatedTags = deduplicateByKey(
			newTagValue === '' ?
				tags.filter(t => t.tag !== tag) :
				tags.map(t => t.tag === tag ? { tag: newTagValue, id: ++nextId } : t),
			'tag'
		);

		sortByKey(updatedTags, 'tag');
		updateItem(itemKey, { tags: updatedTags.map(t => pick(t, ['tag'])) });

		if(tag === '') {
			// if adding a new tag, automatically create input for another one
			updatedTags.push({ tag: '', id: ++nextId });
			setTagRedacted('');
		}

		setTags(updatedTags);
	}

	const handleCancel = () => {
		setTagRedacted(null);
		setTags(tags.filter(t => t.tag !== ''));
	}

	const handleEdit = ev => {
		const tag = ev.currentTarget.closest('[data-tag]').dataset.tag;
		setTagRedacted(tag);
	}

	const handleDelete = ev => {
		const tag = ev.currentTarget.closest('[data-tag]').dataset.tag;
		const updatedTags = tags.filter(t => t.tag !== tag);
		setTags(updatedTags);
		updateItem(itemKey, { tags: updatedTags.map(t => pick(t, ['tag'])) });
	}

	const handleAddTag = () => {
		setTags([...tags.filter(t => t.tag !== ''), { tag: '', id: ++nextId }]);
		setTagRedacted('');
	}

	return (
			<div className="scroll-container-mouse">
				<nav>
					<ul className="details-list tag-list">
						{
							tags.map(tag => {
								return (
									<li className="tag" data-tag={ tag.tag } data-key={ tag.id } key={ tag.id } >
										<Icon
											color={ tag.tag in tagColors ? tagColors[tag.tag] : null }
											height="12"
											symbol={ tag.tag in tagColors ? 'circle' : 'circle-empty' }
											type="12/circle"
											width="12"
										/>
										<Editable
											autoFocus
											isActive={ tag.tag === tagRedacted }
											value={ tag.tag }
											onCommit={ handleCommit }
											onCancel={ handleCancel }
											onClick={ handleEdit }
											onFocus={ handleEdit }
										/>
										{ !isReadOnly && (
											<Button
												icon
												onClick={ handleDelete }
											>
												<Icon type="16/minus-circle" width="16" height="16" />
											</Button>
										)}
									</li>
								);
							})
						}
					</ul>
				</nav>
				{ !isReadOnly && (
					<Toolbar>
						<div className="toolbar-left">
							<ToolGroup>
								<Button
									disabled={ tagRedacted !== null }
									className="btn-link"
									onClick={ handleAddTag }
								>
									<Icon type={ '16/plus' } width="16" height="16" />
									Add Tag
								</Button>
							</ToolGroup>
						</div>
					</Toolbar>
				)}
		</div>
	)
}

Tags.propTypes = {
	isReadOnly: PropTypes.bool,
	itemKey: PropTypes.string,
	tagColors: PropTypes.object,
	tags: PropTypes.array,
	updateItem: PropTypes.func.isRequired,
}

export default Tags;
