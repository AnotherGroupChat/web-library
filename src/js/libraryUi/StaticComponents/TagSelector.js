'use strict';

var log = require('libzotero/lib/Log').Logger('zotero-web-library:TagSelector');

var React = require('react');

var TagSelector = React.createClass({
	getDefaultProps: function() {
		return {};
	},
	getInitialState: function() {
		return {};
	},
	render: function() {
		return (
			<nav id="tag-selector">
				<h3>Tags</h3>
			</nav>
		);
	}
});

module.exports = TagSelector;
