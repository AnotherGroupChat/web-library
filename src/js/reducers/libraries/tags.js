'use strict';

import { get, indexByGeneratedKey } from '../../utils';

import {
    RECEIVE_LIBRARY_SETTINGS,
    RECEIVE_TAGS_IN_COLLECTION,
    RECEIVE_TAGS_IN_LIBRARY,
    RECEIVE_TAGS_FOR_ITEM,
} from '../../constants/actions';

const tags = (state = {}, action) => {
	switch(action.type) {
		case RECEIVE_LIBRARY_SETTINGS:
			return {
				...state,
				...indexByGeneratedKey(
					get(action.settings, 'tagColors.value', []),
					({ name }) => `${name}-0`,
					({ name, ...values }) => ({ tag: name, ...values })
				)
			};
		case RECEIVE_TAGS_IN_LIBRARY:
		case RECEIVE_TAGS_IN_COLLECTION:
		case RECEIVE_TAGS_FOR_ITEM:
			return {
				...state,
				...indexByGeneratedKey(action.tags,
					tag => `${tag.tag}-${tag[Symbol.for('meta')].type}`,
					tag => ({
					...state[`${tag.tag}-${tag[Symbol.for('meta')].type}`],
					...tag
				}))
			}
		default:
			return state;
	}
};

export default tags;
