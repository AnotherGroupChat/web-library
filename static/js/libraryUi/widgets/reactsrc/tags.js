Zotero.ui.widgets.reacttags = {};

Zotero.ui.widgets.reacttags.init = function(el){
	Z.debug("tags widget init", 3);
	var library = Zotero.ui.getAssociatedLibrary(el);

	var reactInstance = ReactDOM.render(
		<Tags library={library} />,
		document.getElementById('tags-list-div')
	);
	Zotero.ui.widgets.reacttags.reactInstance = reactInstance;

	var tagColors = library.preferences.getPref("tagColors");
	var selectedTags = Zotero.state.getUrlVar('tag');
	if(!J.isArray(selectedTags)){
		if(selectedTags) {
			selectedTags = [selectedTags];
		}
		else {
			selectedTags = [];
		}
	}
	
	reactInstance.setState({tagColors: tagColors, selectedTags:selectedTags});
	
	library.listen("tagsDirty", reactInstance.syncTags, {});
	library.listen("cachedDataLoaded", reactInstance.syncTags, {});
	library.listen("tagsChanged libraryTagsUpdated selectedTagsChanged", function(evt){
		reactInstance.setState({tags:library.tags});
	}, {});
	
	var container = J(el);
};

var TagRow = React.createClass({
	getDefaultProps: function(){
		return {
			showAutomatic: false,
		}
	},
	handleClick: function(evt) {
		evt.stopPropagation();
		evt.preventDefault();
		var tag = this.props.tag;
		
		Z.state.toggleTag(tag.apiObj.tag);
		Z.state.clearUrlVars(['tag', 'collectionKey']);
        Z.state.pushState();

        var selectedTags = Zotero.state.getUrlVar('tag');
        if(!J.isArray(selectedTags)){
			if(selectedTags) {
				selectedTags = [selectedTags];
			}
			else {
				selectedTags = [];
			}
		}
		Zotero.ui.widgets.reacttags.reactInstance.setState({selectedTags:selectedTags});
	},
	render: function() {
		var tag = this.props.tag;
		var title = tag.apiObj.tag;
		if(tag.apiObj.meta.numItems) {
			title += " (" + tag.apiObj.meta.numItems + ")";
		}
		var newUrl = "";

		var tagStyle = {};
		if(tag.color){
			tagStyle = {
				color: tag.color,
				fontWeight: "bold",
			}
		}
		
		//render nothing for automatic tags user doesn't want displayed
		if((this.props.showAutomatic == false) && (tag.apiObj.meta.type != 0)){
			return null;
		}

		return (
			<li className="tag-row">
				<a onClick={this.handleClick} className='tag-link' title={title} style={tagStyle} href={newUrl}>{Zotero.ui.trimString(tag.apiObj.tag, 12)}</a>
			</li>
		);
	}
});

var TagList = React.createClass({
	getDefaultProps: function(){
		return {
			tags: [],
			showAutomatic: false,
			id: "",
		}
	},
	render: function() {
		var showAutomatic = this.props.showAutomatic;
		var tagRowNodes = this.props.tags.map(function(tag, ind){
			return (
				<TagRow key={tag.apiObj.tag} tag={tag} showAutomatic={showAutomatic} />
			);
		});

		return (
			<ul id={this.props.id}>
				{tagRowNodes}
			</ul>
		);
	}
});

var Tags = React.createClass({
	getDefaultProps: function() {
		return {}
	},
	getInitialState: function() {
		return {
			tags: new Zotero.Tags(),
			tagColors: null,
			selectedTags: [],
			tagFilter: "",
			showAutomatic: false,
			loading:false,
		};
	},
	handleFilterChanged: function(evt) {
		this.setState({tagFilter: evt.target.value});
	},
	syncTags: function(evt) {
		Z.debug("Tags.syncTags");
		if(this.state.loading){
			return;
		}
		
		var library = this.props.library;

		//clear tags if we're explicitly not using cached tags
		if(evt.data.checkCached === false){
			library.tags.clear();
		}
		
		Zotero.ui.widgets.reacttags.reactInstance.setState({tags:library.tags, loading:true});
		
		//cached tags are preloaded with library if the cache is enabled
		//this function shouldn't be triggered until that has already been done
		loadingPromise = library.loadUpdatedTags()
		.then(function(){
			Z.debug("syncTags done. clearing loading div");
			Zotero.ui.widgets.reacttags.reactInstance.setState({tags:library.tags, loading:false});
			return;
		},
		function(error){
			Z.error("syncTags failed. showing local data and clearing loading div");
			Zotero.ui.widgets.reacttags.reactInstance.setState({tags:library.tags, loading:false});
			Zotero.ui.jsNotificationMessage("There was an error loading tags. Some tags may not have been updated.", 'notice');
		});
		
		return loadingPromise;
	},
	render: function() {
		var tags = this.state.tags;
		var selectedTagStrings = this.state.selectedTags;
		var tagColors = this.state.tagColors;
		if(tagColors === null){
			tagColors = [];
		}

		var matchAnyFilter = this.state.tagFilter;
		var plainTagsList = tags.plainTagsList(tags.tagsArray);
		var matchedTagStrings = Z.utils.matchAnyAutocomplete(matchAnyFilter, plainTagsList);

		var tagColorStrings = [];
		var coloredTags = [];
		tagColors.forEach(function(tagColor, index){
			tagColorStrings.push(tagColor.name.toLowerCase());
			var coloredTag = tags.getTag(tagColor.name);
			if(coloredTag){
				coloredTag.color = tagColor.color;
				coloredTags.push(coloredTag);
			}
		});

		var filteredTags = [];
		var selectedTags = [];

		//always show selected tags, even if they don't pass the filter
		selectedTagStrings.forEach(function(tagString){
			var t = tags.getTag(tagString);
			if(t){
				selectedTags.push(t);
			}
		});
		//add to filteredTags if passes filter, and is not already selected or colored
		matchedTagStrings.forEach(function(matchedString){
			var t = tags.getTag(matchedString);
			if(t !== null && (t.apiObj.meta.numItems > 0)) {
				//we have the actual tag object, and it has items
				//add to filteredTags if it is not already in colored or selected lists
				if(selectedTagStrings.indexOf(matchedString) == -1 && tagColorStrings.indexOf(matchedString) == -1){
					filteredTags.push(t);
				}
			}
		});

		return (
			<div>
				<input type="text" id="tag-filter-input" className="tag-filter-input form-control" placeholder="Filter Tags" onChange={this.handleFilterChanged} />
				<LoadingSpinner loading={this.state.loading} />
				<TagList tags={selectedTags} id="selected-tags-list" />
				<TagList tags={coloredTags} id="colored-tags-list" />
				<TagList tags={filteredTags} id="tags-list" />
			</div>
		);
	}
});
