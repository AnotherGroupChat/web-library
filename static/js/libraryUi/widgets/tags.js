Zotero.ui.widgets.tags = {};

Zotero.ui.widgets.tags.init = function(el){
    Zotero.ui.eventful.listen("tagsDirty", Zotero.ui.widgets.tags.syncTags, {widgetEl: el});
    Zotero.ui.eventful.listen("cachedDataLoaded", Zotero.ui.widgets.tags.syncTags, {widgetEl: el});
    Zotero.ui.eventful.listen("libraryTagsUpdated selectedTagsChanged", Zotero.ui.widgets.tags.rerenderTags, {widgetEl: el});
    
    var container = J(el);
    
    //initialize binds for widget
    //send pref to website when showAllTags is toggled
    container.on('click', "#show-all-tags", Zotero.ui.showAllTags);
    container.on('click', "#show-more-tags-link", Zotero.ui.showMoreTags);
    container.on('click', "#show-fewer-tags-link", Zotero.ui.showFewerTags);
    
    Zotero.ui.bindTagLinks();
    //TODO: make sure tag autocomplete works when editing items
    //add tag to item and stop event propogation when tag is selected
    //from autocomplete on an item
    
    //bind tag autocomplete filter in tag widget
    container.on('keyup', "#tag-filter-input", Zotero.ui.widgets.tags.filterTags);
    
    
    //send pref to website when showAllTags is toggled
    container.on('click', "#show-all-tags", function(e){
        var show = J(this).prop('checked') ? true : false;
        Z.debug("showAllTags is " + show, 4);
        Zotero.utils.setUserPref('library_showAllTags', show);
        Zotero.ui.eventful.trigger('libraryTagsUpdated');
    });
    
    container.on('click', "#show-more-tags-link", function(e){
        e.preventDefault();
        var jel = J(this).closest('#tags-list-div');
        jel.data('showmore', true);
        Zotero.ui.eventful.trigger('libraryTagsUpdated');
    });
    container.on('click', "#show-less-tags-link", function(e){
        e.preventDefault();
        var jel = J(this).closest('#tags-list-div');
        jel.data('showmore', false);
        Zotero.ui.eventful.trigger('libraryTagsUpdated');
    });
    
};

Zotero.ui.widgets.tags.syncTags = function(event){
    Z.debug('Zotero eventful syncTags', 1);
    var widgetEl = J(event.data.widgetEl);
    var checkCached = event.data.checkCached;
    if(checkCached !== false){
        checkCached = true; //default to using the cache
    }
    
    Zotero.state.flagLoading(widgetEl);
    Zotero.ui.showSpinner(widgetEl.find('div.loading'));
    
    //get Zotero.Library object if already bound to element
    var library = Zotero.ui.getAssociatedLibrary(widgetEl);
    //clear tags if we're explicitly not using cached tags
    if(checkCached === false){
        library.tags.clear();
    }
    
    //cached tags are preloaded with library if the cache is enabled
    //this function shouldn't be triggered until that has already been done
    return library.loadUpdatedTags()
    .then(function(){
        Zotero.state.doneLoading(widgetEl);
        widgetEl.find('.loading').empty();
        Zotero.ui.eventful.trigger("libraryTagsUpdated");
    },
    function(){
        //sync failed, but we still have some local data, so show that
        Zotero.ui.eventful.trigger("libraryTagsUpdated");
        Zotero.state.doneLoading(widgetEl);
        widgetEl.children('.loading').empty();
        //TODO: display error as well
    });
};

Zotero.ui.widgets.tags.rerenderTags = function(event){
    Zotero.debug("Zotero eventful rerenderTags", 3);
    var widgetEl = J(event.data.widgetEl);
    
    // put the selected tags into an array
    var selectedTags = Zotero.state.getUrlVar('tag');
    if(!J.isArray(selectedTags)){
        if(selectedTags) {
            selectedTags = [selectedTags];
        }
        else {
            selectedTags = [];
        }
    }
    
    widgetEl.children(".loading").empty();
    var library = Zotero.ui.getAssociatedLibrary(widgetEl);
    var plainList = library.tags.plainTagsList(library.tags.tagsArray);
    Zotero.ui.displayTagsFiltered(widgetEl, library.tags, plainList, selectedTags);
};

//generate html for tags
/**
 * Display filtered list of tags
 * @param  {Dom element} el                 Container
 * @param  {Zotero_Tags} libtags            Zotero_Tags object
 * @param  {array} matchedTagStrings  tags that matched the filter string
 * @param  {array} selectedTagStrings tags that are currently selected
 * @return {undefined}
 */
Zotero.ui.displayTagsFiltered = function(el, libtags, matchedTagStrings, selectedTagStrings){
    Zotero.debug("Zotero.ui.displayTagsFiltered", 3);
    Z.debug(selectedTagStrings, 4);
    var curPreString = J("#tag-filter-input").val();
    var jel = J(el);
    var showMore = jel.data('showmore');
    if(!showMore){
        showMore = false;
    }
    
    var filteredTags = [];
    var selectedTags = [];
    J.each(matchedTagStrings, function(index, matchedString){
        if(libtags.tagObjects[matchedString] && (J.inArray(matchedString, selectedTagStrings) == (-1))) {
            filteredTags.push(libtags.tagObjects[matchedString]);
        }
    });
    J.each(selectedTagStrings, function(index, selectedString){
        if(libtags.tagObjects[selectedString]){
            selectedTags.push(libtags.tagObjects[selectedString]);
        }
    });
    
    var passTags;
    if(!showMore){
        passTags = filteredTags.slice(0, 25);
        J("#show-more-tags-link").show();
        J("#show-fewer-tags-link").hide();
    }
    else{
        passTags = filteredTags;
        J("#show-more-tags-link").hide();
        J("#show-fewer-tags-link").show();
    }
    
    var tagListEl = J("#tags-list").empty();
    J("#selected-tags-list").replaceWith(J('#tagunorderedlistTemplate').render({tags:selectedTags, id:'selected-tags-list'}));
    J("#tags-list").replaceWith(J('#tagunorderedlistTemplate').render({tags:passTags, id:'tags-list'}));
    
};

Zotero.ui.showAllTags = function(e){
    var show = J(this).prop('checked') ? true : false;
    Z.debug("showAllTags is " + show, 4);
    Zotero.utils.setUserPref('library_showAllTags', show);
    Zotero.ui.eventful.trigger('libraryTagsUpdated');
};

Zotero.ui.showMoreTags = function(e){
    e.preventDefault();
    var jel = J(this).closest('#tags-list-div');
    jel.data('showmore', true);
    Zotero.ui.eventful.trigger('libraryTagsUpdated');
};

Zotero.ui.showFewerTags = function(e){
    e.preventDefault();
    var jel = J(this).closest('#tags-list-div');
    jel.data('showmore', false);
    Zotero.ui.eventful.trigger('libraryTagsUpdated');
};


/**
 * Bind tag links to take appropriate action rather than following the link
 * @return {undefined}
 */
Zotero.ui.bindTagLinks = function(container){
    Z.debug("Zotero.ui.bindTagLinks", 3);
    J(container).on('click', 'a.tag-link', function(e){
        e.preventDefault();
        J("#tag-filter-input").val('');
        var tagtitle = J(this).attr('data-tagtitle');
        Zotero.state.toggleTag(tagtitle);
        Zotero.state.clearUrlVars(['tag', 'collectionKey']);
        Zotero.state.pushState();
    });
};

Zotero.ui.widgets.tags.filterTags = function(e){
    Z.debug("Zotero.ui.widgets.tags.filterTags", 3);
    var library = Zotero.ui.getAssociatedLibrary(J('#tag-filter-input').closest('.eventfulwidget'));
    var libraryTagsPlainList = library.tags.plainList;
    var matchingTagStrings = Zotero.utils.matchAnyAutocomplete(J('#tag-filter-input').val(), libraryTagsPlainList);
    Zotero.ui.displayTagsFiltered(J('#tags-list-div'), library.tags, matchingTagStrings, []);
    Z.debug(matchingTagStrings, 4);
};

