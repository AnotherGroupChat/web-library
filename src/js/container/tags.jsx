'use strict';

import React from 'react';
import { connect } from 'react-redux';
import { updateItem } from '../actions/';
import { get } from '../utils';
import Tags from '../component/tags';

const TagsContainer = props => <Tags { ...props } />;

const mapStateToProps = state => {
	const { libraryKey, itemKey } = state.current;
	const { tags } = get(state, ['libraries', libraryKey, 'items', itemKey], {});
	const tagColors = get(state, ['libraries', libraryKey, 'tagColors']);
	return { libraryKey, itemKey, tagColors, tags };
}

export default connect(mapStateToProps, { updateItem })(TagsContainer)
