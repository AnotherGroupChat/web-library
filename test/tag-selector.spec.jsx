import React from 'react';
import renderer from 'react-test-renderer';

import TagSelector from '../src/js/component/tag-selector';
import tags from './fixtures/tags.json';

const mapCollectionsRecursive = (col, f) => {
	return col.map(c => {
		c = f(c);
		if(c.children.length) {
			c.children = mapCollectionsRecursive(c.children, f);
		}
		return c;
	});
};

it('Renders tag selector with multiple tags', () => {
	let tree = renderer.create(
		<TagSelector tags = { tags } />
	).toJSON();
	expect(tree).toMatchSnapshot();
});

it('Renders tag selector with a search string', () => {
	let tree = renderer.create(
		<TagSelector tags = { tags.filter(t => t.name.includes('tag')) } searchString="tag" />
	).toJSON();
	expect(tree).toMatchSnapshot();
});
