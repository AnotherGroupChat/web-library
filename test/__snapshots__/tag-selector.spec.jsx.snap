// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders tag selector with a search string 1`] = `
<div
  className="tag-selector"
>
  <div
    className="tag-selector-container"
  >
    <ul
      className="tag-selector-list"
    >
      <li
        className="tag-selector-item colored"
        onClick={[Function]}
        onContextMenu={[Function]}
        style={
          Object {
            "color": "#ff9900",
          }
        }
      >
        tag1
      </li>
      <li
        className="tag-selector-item colored"
        onClick={[Function]}
        onContextMenu={[Function]}
        style={
          Object {
            "color": "rgba(55, 255, 55, 1)",
          }
        }
      >
        tag2
      </li>
    </ul>
  </div>
  <div
    className="tag-selector-filter-container"
  >
    <input
      className="tag-selector-filter"
      onChange={[Function]}
      type="search"
      value="tag"
    />
    <button
      className="tag-selector-actions"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`Renders tag selector with multiple tags 1`] = `
<div
  className="tag-selector"
>
  <div
    className="tag-selector-container"
  >
    <ul
      className="tag-selector-list"
    >
      <li
        className="tag-selector-item colored"
        onClick={[Function]}
        onContextMenu={[Function]}
        style={
          Object {
            "color": "#ff9900",
          }
        }
      >
        tag1
      </li>
      <li
        className="tag-selector-item colored"
        onClick={[Function]}
        onContextMenu={[Function]}
        style={
          Object {
            "color": "rgba(55, 255, 55, 1)",
          }
        }
      >
        tag2
      </li>
      <li
        className="tag-selector-item selected colored"
        onClick={[Function]}
        onContextMenu={[Function]}
        style={
          Object {
            "color": "rgb(55, 55, 255)",
          }
        }
      >
        lorem
      </li>
      <li
        className="tag-selector-item selected colored"
        onClick={[Function]}
        onContextMenu={[Function]}
        style={
          Object {
            "color": "red",
          }
        }
      >
        ipsum
      </li>
      <li
        className="tag-selector-item"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        foo
      </li>
      <li
        className="tag-selector-item"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        bar
      </li>
      <li
        className="tag-selector-item"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        uho
      </li>
      <li
        className="tag-selector-item"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        mehe
      </li>
      <li
        className="tag-selector-item"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        ala
      </li>
      <li
        className="tag-selector-item disabled"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        wolf
      </li>
      <li
        className="tag-selector-item disabled"
        onClick={[Function]}
        onContextMenu={[Function]}
      >
        prolf
      </li>
    </ul>
  </div>
  <div
    className="tag-selector-filter-container"
  >
    <input
      className="tag-selector-filter"
      onChange={[Function]}
      type="search"
      value=""
    />
    <button
      className="tag-selector-actions"
      onClick={[Function]}
    />
  </div>
</div>
`;
